streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/DefaultStateConstrainedBalancedAssignor.java

```diff
@@ -0,0 +1,218 @@
+            String clientWithLeastTasks = null;
+            int taskCount = Integer.MAX_VALUE;
+            for (final String client : caughtUpClients) {
+                final int assignedTasksCount = assignment.get(client).size();
+                if (taskCount > assignedTasksCount) {
+                    clientWithLeastTasks = client;
+                    taskCount = assignedTasksCount;
+                }
+            }
+            assignment.get(clientWithLeastTasks).add(taskId);
+        }
+    }
+
+    /**
+     * Assigns tasks for which no caught-up clients exist.
+     * A task is assigned to one of the clients with the least lag and the least tasks assigned.
+     *
+     * @param assignment assignment
+     * @param unassignedTasksWithoutCaughtUpClients list of task that could not be assigned since no caught-up client exists for them
+     * @param statefulTasksToRankedClients ranked clients map
+     */
+    private void assignTasksWithoutCaughtUpClients(final Map<String, List<TaskId>> assignment,
+                                                   final List<TaskId> unassignedTasksWithoutCaughtUpClients,
+                                                   final Map<TaskId, List<ClientIdAndLag>> statefulTasksToRankedClients) {
+        for (final TaskId taskId : unassignedTasksWithoutCaughtUpClients) {
+            List<ClientIdAndLag> clientIdAndLags = statefulTasksToRankedClients.get(taskId);
+            int previousLag = clientIdAndLags.get(0).lag();
+            String clientWithLeastTasks = null;
+            int taskCount = Integer.MAX_VALUE;
+            int i = 0;
+            while (clientIdAndLags.get(i).lag() == previousLag) {
+                final String clientId = clientIdAndLags.get(i).clientId();
+                final int assignedTasksCount = assignment.get(clientId).size();
+                if (taskCount > assignedTasksCount) {
+                    clientWithLeastTasks = clientId;
+                    taskCount = assignedTasksCount;
+                }
+                ++i;
+            }
+            assignment.get(clientWithLeastTasks).add(taskId);
+        }
+    }
+
+    /**
+     * Balance the assignment.
+     *
+     * @param assignment assignment
+     * @param balanceFactor balance factor
+     * @param statefulTasksToRankedClients ranked clients map
+     * @param tasksToCaughtUpClients map from task IDs to lists of caught-up clients
+     */
+    private void balance(final Map<String, List<TaskId>> assignment,
+                         final int balanceFactor,
+                         final Map<TaskId, List<ClientIdAndLag>> statefulTasksToRankedClients,
+                         final Map<TaskId, List<String>> tasksToCaughtUpClients) {
+        final List<String> clients = new ArrayList<>(assignment.keySet());
+        if (clients.isEmpty()) {
+            throw new IllegalStateException("Clients within an assignment must not be empty");
```

Comment:
req: If `clients` is empty I think we can/should catch it earlier, like back in `initAssignment`?